clarity/src/vm/types/signatures.rs

```diff
@@ -641,13 +641,26 @@ impl TypeSignature {
                 }
             }
             NoType => Err(CheckErrors::CouldNotDetermineType),
-            TraitReferenceType(_) => {
-                unreachable!("TraitReferenceType should not be used in epoch v2.1")
-            }
             _ => Ok(&other == self),
         }
     }
 
+    /// Canonicalize the type. This method will replace any types from previous
+    /// epochs with the appropriate types for the current epoch.
+    pub fn canonicalize(&self, epoch: &StacksEpochId) -> TypeSignature {
+        match epoch {
+            StacksEpochId::Epoch21 => self.canonicalize_v2_1(),
+            _ => self.clone(),
+        }
+    }
+
+    pub fn canonicalize_v2_1(&self) -> TypeSignature {
+        match self {
+            TraitReferenceType(trait_id) => CallableType(CallableSubtype::Trait(trait_id.clone())),
+            _ => self.clone(),
+        }
+    }
+
```

Comment:
Does this work if the TraitReferenceType is nested in an optional?

Reply:
It would get handled, but only because the function `canonicalize` is called from (`clarity2_inner_type_check_type`) is recursing into the type. Would it be more clear if this method itself recursed into the type and was called somewhere else? Or maybe if this function was just renamed to `canonicalize_simple_type` or canonicalize_base_type`?

Reply:
Yeah -- I think just renaming and adding a note to the rustdocs would be sufficient here. As it's currently used, it's fine, we should just make sure that a future caller doesn't think that it will canonicalize a complex type.

Reply:
Thanks. Added in f5567b216.